### ISSUE TITLE
Rule proposal: `default-export-function-style`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#1089]
Whether to prefer a default export as a named function or a const variable with an array function.

It should default to `function` style.

It should be auto-fixable.

## Fail

```js
// "unicorn/default-export-function-style": ["error", "function"]

const foo = () => {};
export default foo;
```

```js
// "unicorn/default-export-function-style": ["error", "variable"]

export default function foo() {}
```

## Pass

```js
// "unicorn/default-export-function-style": ["error", "function"]

export default function foo() {}
```

```js
// "unicorn/default-export-function-style": ["error", "variable"]

const foo = () => {};
export default foo;
```

**Comment by fregante** [comment]
> Open question. What should happen to:

I think that that code won’t pass an existing rule already (default exports should be named)

**Comment by sindresorhus** [comment]
> I think that that code won’t pass an existing rule already: https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-anonymous-default-export.md

I'm aware, but we cannot assume other rules. We need to decide how to handle that case.

**Comment by fregante** [comment]
Then I think it should be covered by the “function” style

**Comment by fregante** [comment]
I don’t think it should be called "function/variable" however, they're all functions:

```js
function foo() {}
const foo = () => {}
let foo = function () {}
```

Maybe these are `inline`:

```js
export default function () {}
export default function foo () {}
export default () => {}
```

And this should be `pre-defined` or something:

```js
export default foo; // However the `foo` function is defined is irrelevant
```

---

Discussing this rule further:

- Do you want to apply the same rule to classes? `export default class {}`? Objects? Where does it stop?

**Comment by sindresorhus** [comment]
Sounds good. How about `statement` and `inline`?

**Comment by sindresorhus** [comment]
> Do you want to apply the same rule to classes? export default class {}? Objects? Where does it stop? export default 4 instead of const n = 4; export default n;?

I think it makes sense for classes, but not objects or variables.

Maybe we should just do an options-object?

```js
{
	functions: 'statement',
	classes: 'statement',
	objects: 'inline',
	values: 'inline'
}
```

**Comment by fregante** [comment]
If we agree with `import/no-anonymous-default-export` I don’t think it’s worth your time to implement something you’re not going to use (`objects: inline`, for example), but that’s up to you.

**Comment by sindresorhus** [comment]
> I’d limit it to classes and functions and a single configuration string: ['error', 'inline'/'statement']

Alright. Let's try that first.

**Comment by fisker** [comment]
Personally I don't use `inline` (except some really simple functions), what style do you use? @fregante @sindresorhus

**Comment by fregante** [comment]
I use inline exports all the time. For defaults, I still prefer inline exports if they can be named inline (functions, classes)

**Comment by sindresorhus** [comment]
I used to prefer non-inline so I could use arrow functions, but lately I'm preferring inline `export default function foo () {}`.

**Comment by fisker** [comment]
I only care function/class is anonymous or not. I'm fine with

```js
class A {}
export default A;
```

```js
export default class A {};
```

```js
function a {}
export default a;
```

```js
export default function a {}
```

```js
export default [];
export default {};
export default 1;
```

But not

```js
export default function () {}
```

```js
export default class () {}
```

```js
export default () => {}
```

**Comment by fregante** [comment]
That rule already exists, I already don't like anonymous exports.

**Comment by fisker** [comment]
I know, but I don't want all [these options](https://github.com/prettier/prettier/blob/ed23dacc9e655c3876971b30859497b17ff2cf9f/eslint.config.js#L155-L164)

**Comment by fisker** [comment]
Maybe we can also suggest a fix based on the filename.
